### PR TITLE
dr/fix-forced-rupture-time

### PR DESCRIPTION
**Pull request by sebwolf-de**
Rather big refactoring, ~will add tpv16 as a test to the CI in another PR~.

### CODE REVIEW

**Comment by ravil-mobile** [comment]
I guess we don't need `tn` anymore, right?
https://github.com/SeisSol/SeisSol/blob/f688be461f3bf710aabe1768f33998d15212ca58/src/Initializer/DynamicRupture.h#L175-L177

File: src/Initializer/DynamicRupture.h [hunk 1]
```diff
@@ -174,13 +174,11 @@ struct seissol::initializers::LTS_LinearSlipWeakening : public seissol::initiali
 
 struct seissol::initializers::LTS_LinearSlipWeakeningForcedRuptureTime : public seissol::initializers::LTS_LinearSlipWeakening {
   Variable<real[dr::misc::numPaddedPoints]> forcedRuptureTime;
-  Variable<real> tn;
```
Comment (sebwolf-de) [inline src/Initializer/DynamicRupture.h:177]: @ravil-mobile I removed this parameter here.

File: src/DynamicRupture/FrictionLaws/LinearSlipWeakening.cpp [hunk 2]
```diff
@@ -1,142 +1,64 @@
...
+  // modify strength according to prakash clifton
+  // see e.g.: Pelties - Verification of an ADER-DG method for complex dynamic rupture problems
+  prak_clif_mod(strength, sigma, localSlipRate, mu, deltaT);
+  // save for output
+  regularisedStrength[ltsFace][pointIndex] = strength;
+}
+
 /*
- * calculates strength
+ * calculates prakash clifton regularization
```
Comment (Thomas-Ulrich) [inline src/DynamicRupture/FrictionLaws/LinearSlipWeakening.cpp:59]: Prakash-Clifton

File: src/DynamicRupture/FrictionLaws/LinearSlipWeakening.cpp [hunk 3]
```diff
@@ -1,142 +1,64 @@
...
+                                   real& sigma,
+                                   real& mu,
+                                   real& deltaT,
+                                   unsigned int ltsFace,
+                                   unsigned int pointIndex) {
+  // modify strength according to prakash clifton
+  // see e.g.: Pelties - Verification of an ADER-DG method for complex dynamic rupture problems
+  prak_clif_mod(strength, sigma, localSlipRate, mu, deltaT);
+  // save for output
+  regularisedStrength[ltsFace][pointIndex] = strength;
```
Comment (Thomas-Ulrich) [inline src/DynamicRupture/FrictionLaws/LinearSlipWeakening.cpp:55]: why not returning the regularized strength in prak_clif_mod?

File: src/DynamicRupture/FrictionLaws/LinearSlipWeakening.cpp [hunk 4]
```diff
@@ -1,142 +1,64 @@
...
 }
 
+void BiMaterialFault::strengthHook(real& strength,
+                                   real& localSlipRate,
+                                   real& sigma,
+                                   real& mu,
+                                   real& deltaT,
+                                   unsigned int ltsFace,
+                                   unsigned int pointIndex) {
+  // modify strength according to prakash clifton
```
Comment (Thomas-Ulrich) [inline src/DynamicRupture/FrictionLaws/LinearSlipWeakening.cpp:51]: Prakash-Clifton

File: src/DynamicRupture/FrictionLaws/LinearSlipWeakening.h [hunk 5]
```diff
@@ -139,20 +148,53 @@ class LinearSlipWeakeningBase : public BaseFrictionLaw<LinearSlipWeakeningBase<D
...
+    // the reference triangle with degree less or equal than CONVERGENCE_ORDER-1, and then evaluates
+    // the polynomial at the quadrature points
+    resampleKrnl.execute();
+
+    for (unsigned pointIndex = 0; pointIndex < misc::numPaddedPoints; pointIndex++) {
+      // integrate slip rate to get slip = state variable
+      this->accumulatedSlipMagnitude[ltsFace][pointIndex] +=
+          resampledSlipRate[pointIndex] * this->deltaT[timeIndex];
+
+      // Modif T. Ulrich-> generalisation of tpv16/17 to 30/31
```
Comment (Thomas-Ulrich) [inline src/DynamicRupture/FrictionLaws/LinearSlipWeakening.h:189]: you can remove this line

**Review by Thomas-Ulrich — APPROVED** [review]
I wonder if having a specialiazation for ForcedRuptureTime is really worth it, I mean, can't we just keep this feature for all LinearSlipWeakeningLaws?
tpv6 benchmark does not have ForcedRuptureTime, but it would make sense to be able to use bimaterial with ForcedRuptureTime.
(then we would initialize ForcedRuptureTime with a very large number to disable it).

**Comment by sebwolf-de** [comment]
> I wonder if having a specialiazation for ForcedRuptureTime is really worth it, I mean, can't we just keep this feature for all LinearSlipWeakeningLaws?

Sure, we could do that. At the same time, we have to include the nucleation stress option, which was recently added to the master branch. How does ForcedRuptureTime interact with the NucleationStress option?
Now, I see four options: `{LSW with ForcedRuptureTime, LSW with NucleationStress} X {LSW with Prakash-Clifton, LSW without Prakash-Clifton}`. Is that right? Standard behaviour (e.g. for tpv5) can be achieved by using `(ForcedRuptureTime, No Prakash-Clifton)` and `ForcedRuptureTime = +inf`.

**Comment by Thomas-Ulrich** [comment]
>Now, I see four options: {LSW with ForcedRuptureTime, LSW with NucleationStress} X {LSW with Prakash-Clifton, LSW without Prakash-Clifton}.
I think so too.

**Comment by sebwolf-de** [comment]
I made Forced Rupture Time the default now and deprecated FL16. Still, if no `forced_rupture_time` is given in `fault.yaml`, I set it to a big number, so we have the same behaviour as for FL2.
I would be very happy, if we could leave out Prakash-Clifton for this PR and take care of that in another PR.

**Review by Thomas-Ulrich — APPROVED** [review]
Thanks for implementing that!

**Pull request merged**